docs/coding_guidelines.md

```diff
@@ -0,0 +1,51 @@
+# Coding guidelines
+
+The goal of this document is to establish some common understanding of what
+language features should and should not be used. This helps reduce mental
+overhead while reading, developing, and reviewing code. In some cases, it
+also helps present a more consistent user experience.
+
+Discussions about this document should occur in a pull request making changes
+to the text of this document. This helps keep track of why things are the way
+they are. As well as providing a structured environment to participate.
+
+If something is not called out in this document, defer to [the C++ core
+guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines).
+In other words, this document supersedes the core guidelines.
+
+Note that the checked in code may not adhere to these guidelines yet.
+This document is intended to be referenced for new code. Ideally we also
+modify checked-in code to follow these guidelines (if/when there is time).
+
+## Error handling and exceptions
```

Comment:
Should we also mention how and when to use the `LOG(ERROR)` facility?

Reply:
Do you mean in context of AST passes?

Reply:
Well, we're using it in other places, too. But yes, in context of AST passes it's especially useful as using it makes the pass fail.